### ISSUE TITLE
lstm related tests are failing due to unprovided optional inputs are not read correctly as nullptr on s390x machine

### ISSUE THREAD

**Issue by Sidong-Wei** [tensorflow/tensorflow#45556]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): N
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8.5
- Bazel version (if compiling from source): 3.1.0
- GCC/Compiler version (if compiling from source): gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
`NoCifgNoPeepholeNoProjectionNoClippingLstmTest.LstmBlackBoxTest` from `nnapi_delegate_test` is failing due to seg fault on s390x. I looked into this issue a bit and found that it has this error because of this flag from `tensorflow/lite/kernels/lstm_eval.cc` is not set properly when `use_layer_norm` should be `false`.
```
const bool use_layer_norm = (forget_layer_norm_coefficients_ptr != nullptr);
```

As I checked further, I found that this is an interesting issue because these are the 4 optional inputs that come together:
```
// Layer norm coefficients of size 'n_cell', representing diagonal matrices.
//   input_layer_norm_coefficients_ptr  - optional
//   forget_layer_norm_coefficients_ptr - optional
//   cell_layer_norm_coefficients_ptr   - optional
//   output_layer_norm_coefficients_ptr - optional
```
So it looks like they should be either all be used or all not be used. However, I found that when the test model does not have these 4 optional inputs, only one of them is `nullptr` while the other 3 are some random values. On an x86 machine this one input is `forget_layer_norm_coefficients_ptr` while on an s390x machine this one input is `cell_layer_norm_coefficients_ptr`. In other words, if I change this flag check into:
```
const bool use_layer_norm = (cell_layer_norm_coefficients_ptr != nullptr);
```
then the test will pass on s390x and fail with seg fault on x86. 
I just wonder is this flag check just randomly choose `forget_layer_norm_coefficients_ptr` or there are some rational behind it? It looks to me that the 4 optional inputs are always treated equally in the `LSTMOpModel` object in the test file, but there is this subtle difference between them. It will be very helpful to understand why only `forget_layer_norm_coefficients_ptr` is picked here so that I will be able to fix it for s390x rather than simply modifying the flag definition.

After solving this issue, another lstm test case `NoCifgPeepholeProjectionClippingLstmTest.LstmBlackBoxTest` fails again in this test file. This time it is due to the `GetTensorData` function in `tensorflow/lite/kernels/internal/tensor_ctypes.h` does not return a `nullptr` for tensor of type `kTfLiteNoType`. I think all the tensor of type `kTfLiteNoType` should have it `data.raw` field points to a `nullptr`, but looks it is breaking for this test case only. I have to make the following change to fix it:
```diff
diff --git a/tensorflow/lite/kernels/internal/tensor_ctypes.h b/tensorflow/lite/kernels/internal/tensor_ctypes.h
index f1d3e17fcb..f5165cc0f0 100644
--- a/tensorflow/lite/kernels/internal/tensor_ctypes.h
+++ b/tensorflow/lite/kernels/internal/tensor_ctypes.h
@@ -27,8 +27,8 @@ inline T* GetTensorData(TfLiteTensor* tensor) {
 template <typename T>
 inline const T* GetTensorData(const TfLiteTensor* tensor) {
-  return tensor != nullptr ? reinterpret_cast<const T*>(tensor->data.raw)
-                           : nullptr;
+  return tensor != nullptr ? (tensor->type != kTfLiteNoType ?
+                 reinterpret_cast<const T*>(tensor->data.raw) : nullptr) : nullptr;
 }
 inline RuntimeShape GetTensorShape(const TfLiteTensor* tensor) {
```
I am not sure but it looks like these two issues are related. They all seem to be originated from the fact that the unprovided optional input is not set correctly. I am also suspecting that it has something to do with the model field initialization process or `lstm.cc` evaluation process, but all these errors are not straightforward so I would appreciate if any help could be provided, thanks.

**Describe the expected behavior**
The unprovided optional input should be `nullptr`, and test case should pass

**Standalone code to reproduce the issue**
Running the test case on s390x will reproduce the error.
```
bazel test --host_javabase="@local_jdk//:jdk" --cache_test_results=no --build_tests_only --test_output=errors -- //tensorflow/lite/delegates/nnapi:nnapi_delegate_test
```

**Other info / logs**

**Comment by abattery** [comment]
@renjie-liu could you take a look at this?

**Comment by renjie-liu** [comment]
It looks like related to the nnapi_delegate_test setup? Miao, can you help take a look?

**Comment by Sidong-Wei** [comment]
To add on a bit more info, there are two other related test: `//tensorflow/lite/kernels:lstm_test` and `//tensorflow/lite/kernels:lstm_eval_test`.
Before any changes were made, `lstm_test` is failing due to seg fault on both my x86 machine and s390x machine. The first fix about `use_layer_norm` is actually making this test proceed further and then fail again.
Before any changes were made, `lstm_eval_test` will pass on both x86 and s390x. The second change I made about `GetTensorData` is actually breaking this test.
It looks to me that these tests are correlated and share some common issues here, still investigating.

**Comment by lev-prol** [comment]
I haven't managed to reproduce the issue yet, but I can shed some light on why `forget_layer_norm_coefficients_ptr` is checked. The reason is because when the cell is used in the CIFG (coupled input and forget gates) mode, input layer norm coefficients will not be present but the rest of the layer norm weights still could be set. As the layer norm weights are either all present or all omitted, `forget_layer_norm_coefficients_ptr` is used as an indicator for all these 3 or 4 tensors.

Also, I also don't think that the problem is in the test setup since the test is [not adding the layer normalisation weights at all](https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/lite/delegates/nnapi/nnapi_delegate_test.cc#L3357-L3384). So the test should use the legacy path where the op has 20 inputs and do not touch layer norm coefficients after `use_layer_norm` is [set to false](https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/lite/kernels/lstm.cc#L1157).

**Comment by Sidong-Wei** [comment]
Hi @lev-prol, thanks a lot for the explanation! If I understand correctly, I think theoretically, in this particular nnapi test case where only 20 inputs are provided, all the 4 layer norm coefficient input should be `nullptr` and we could check `forget_layer_norm_coefficients_ptr` to indicate the rest ones.
However, as I mentioned, if I change `const bool use_layer_norm = (forget_layer_norm_coefficients_ptr != nullptr);` into `const bool use_layer_norm = (cell_layer_norm_coefficients_ptr != nullptr);` at [lstm_eval.cc](https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/lite/kernels/lstm_eval.cc#L226), the test will fail on my x86 machine because `use_layer_norm` became true while it should be false. But I think theoretically, making this change should  not make it fail, is that right?

I have some further info here when running this test case on my x86 machine, when the model [invokes](https://github.com/tensorflow/tensorflow/blob/fcc4b966f1265f466e82617020af93670141b009/tensorflow/lite/delegates/nnapi/nnapi_delegate_test.cc#L3281) itself and execute until this [line](https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/lite/core/subgraph.cc#L965), I pull out some info into gdb:
```
(gdb) p *node.inputs
$1 = {size = 20, data = 0x55555567beb4}
(gdb) p node.inputs->data[20]
$2 = 0
(gdb) p node.inputs->data[21]
$3 = 33
(gdb) p node.inputs->data[22]
$4 = 0
(gdb) p node.inputs->data[23]
$5 = 1
```
These "out-of-range" input will still be read as the 4 optional input later (because in `eval` [function](https://github.com/tensorflow/tensorflow/blob/fcc4b966f1265f466e82617020af93670141b009/tensorflow/lite/kernels/lstm.cc#L1518) the input is read regardless of its size), and you could notice only input 21 (kForgetLayerNormCoefficientsTensor) points to an invalid tensor index which will be retrieved as a `nullptr` later, the other 3 all points to a valid tensor index so their pointer will not be `nullptr`. So this should be the reason why the other 3 pointers are not `nullptr`, and I am very curious if this could be explained.
(In comparison, this is the info I got for the same test case on s390x machine:)
```
(gdb) p *node.inputs
$3 = {size = 20, data = 0x2aa001194e4}
(gdb) p node.inputs->data[20]
$4 = 0
(gdb) p node.inputs->data[21]
$5 = 0
(gdb) p node.inputs->data[22]
$6 = 33
(gdb) p node.inputs->data[23]
$7 = 1
```

As for the `GetTensorData` issue, my fix seems to cause regression for other test cases but I found this [fix](https://github.com/tensorflow/tensorflow/commit/1970c2158b1ffa416d159d03c3370b9a462aee35) is merged already, so I guess it will be fixed in 2.4.0 release.

**Comment by Sidong-Wei** [comment]
Test case pass on latest release, closing the issue.
